File: clients/js/src/features/config_plugin_instance.ts
import { Input } from "../input";

import { PolywrapClientConfigBuilder, PolywrapClient } from "@polywrap/client-js";
import { PluginWrapper, PluginModule } from "@polywrap/plugin-js";

export async function runTestCase(input: unknown): Promise<void> {
  const inputObj = Input.expectObject<{
    uri: unknown;
    method: unknown;
  }>(input);
  const uri = Input.expectUri(inputObj.uri).uri;
  const method = Input.expectString(inputObj.method);

  console.log("Creating Plugin Instance");

  class Plugin extends PluginModule<{}, {}> {
    counter: number = 0;

    increment() {
      ++this.counter;
    }
  }

  const plugin = new Plugin({});

  console.log("Adding Plugin Instance to ClientConfig");

  const pluginWrapper = new PluginWrapper(
    { version: "0.1", type: "plugin", name: "counter", abi: {} },
    plugin
  );

  const config = new PolywrapClientConfigBuilder()
    .setWrapper(uri, pluginWrapper)
    .build();

  const client = new PolywrapClient(config);

  for (let i = 0; i < 2; ++i) {
    console.log("Invoking Plugin Instance");

    const result = await client.invoke({
      uri,
      method
    });

    if (result.ok) {
      console.log(`counter = ${plugin.counter}`);
    }
  }

  console.log("Success!");
}

File: clients/py/src/features/config_plugin_instance.py
from typing import Any
from polywrap_client import PolywrapClient
from polywrap_client_config_builder import PolywrapClientConfigBuilder
from polywrap_plugin import PluginModule, PluginWrapper

from pydantic import BaseModel

from validators import UriStr


class TestCaseInput(BaseModel):
    uri: UriStr
    method: str


class Plugin(PluginModule[None]):
    counter = 0

    def __init__(self, config: None):
        super().__init__(config)

    def increment(self, *_: Any):
        self.counter += 1


def run_test_case(input: Any) -> None:
    input_obj = TestCaseInput.validate(input)

    print("Creating Plugin Instance")

    plugin = Plugin(None)

    print("Adding Plugin Instance to ClientConfig")

    plugin_wrapper = PluginWrapper(plugin, NotImplemented)

    config = (
        PolywrapClientConfigBuilder().set_wrapper(input_obj.uri, plugin_wrapper).build()
    )

    client = PolywrapClient(config)

    for _ in range(2):
        print("Invoking Plugin Instance")

        try:
            client.invoke(uri=input_obj.uri, method=input_obj.method)

            print(f"counter = {plugin.counter}")
        except Exception as e:
            print(e)

    print("Success!")
